Implement torch.tril_indices and torch.triu_indices (#12653)

mrshenli opened the pull request:
This is an optimized implementation that does the following:

1. created an empty Tensor of correct size.
2. fill the Tensor with correct values.

The following three designs to fill in the Tensor result in roughly the same performance. Hence, the 2nd option is taken for simpler code, and to return contiguous tensors.

1. Sequential: fill row coordinates first, then columns. This results in two for-loop and more arithmetic operations.
2. Interleaved: fill in index coordinates one by one, which jumps between the two output Tensor rows in every iteration.
3. Transpose: create a n X 2 Tensor, fill the Tensor sequentially, and then transpose it.

<img width="352" alt="screen shot 2018-12-10 at 3 54 39 pm" src="https://user-images.githubusercontent.com/16999635/49769172-07bd3580-fc94-11e8-8164-41839185e9f9.png">


NOTE:

This implementation returns a 2D tensor, instead of a tuple of two tensors. It means that users will not be able to do the following:

```python
x = torch.ones(3, 3)
i = torch.tril_indices(3, 3)
x[i]  # need to first convert the 2D tensor into a tuple of two 1D tensors.
```

mrshenli committed:
Implement torch.tril_indices (#12653)

This is a straightforward implementation that uses
  torch.ones(rows, cols).tril(offset).nonzero()

mrshenli committed:
Optimized implementation of tril_indices:

1. created an empty Tensor of correct size.
2. fill the Tensor with correct values.

mrshenli committed:
Support multiple dtypes and device types.

mrshenli committed:
Implement torch.triu_indices

mrshenli commented:
The failure seems to be irrelevant.

```
Dec 09 01:15:26 self = <caffe2.python.onnx.tests.c2_ref_test.TestCaffe2End2End testMethod=test_alexnet>
Dec 09 01:15:26 model = 'bvlc_alexnet'
Dec 09 01:15:26 
Dec 09 01:15:26     def _download(self, model):
Dec 09 01:15:26         model_dir = self._model_dir(model)
Dec 09 01:15:26         assert not os.path.exists(model_dir)
Dec 09 01:15:26         os.makedirs(model_dir)
Dec 09 01:15:26         for f in ['predict_net.pb', 'init_net.pb', 'value_info.json']:
Dec 09 01:15:26             url = getURLFromName(model, f)
Dec 09 01:15:26             dest = os.path.join(model_dir, f)
Dec 09 01:15:26             try:
Dec 09 01:15:26                 try:
Dec 09 01:15:26                     downloadFromURLToFile(url, dest,
Dec 09 01:15:26                                           show_progress=False)
Dec 09 01:15:26                 except TypeError:
Dec 09 01:15:26                     # show_progress not supported prior to
Dec 09 01:15:26                     # Caffe2 78c014e752a374d905ecfb465d44fa16e02a28f1
Dec 09 01:15:26                     # (Sep 17, 2017)
Dec 09 01:15:26                     downloadFromURLToFile(url, dest)
Dec 09 01:15:26             except Exception as e:
Dec 09 01:15:26                 print("Abort: {reason}".format(reason=e))
Dec 09 01:15:26                 print("Cleaning up...")
Dec 09 01:15:26                 deleteDirectory(model_dir)
Dec 09 01:15:26 >               raise AssertionError("Test model downloading failed")
Dec 09 01:15:26 E               AssertionError: Test model downloading failed
Dec 09 01:15:26 
Dec 09 01:15:26 /usr/local/caffe2/lib/python2.7/dist-packages/caffe2/python/onnx/tests/test_utils.py:57: AssertionError
Dec 09 01:15:26 ----------------------------- Captured stdout call -----------------------------
Dec 09 01:15:26 Downloading from https://s3.amazonaws.com/download.caffe2.ai/models/bvlc_alexnet/predict_net.pb
Dec 09 01:15:26 Abort: Could not download model. [URL Error] [Errno 104] Connection reset by peer.
```

zou3519 commented:
@mrshenli that seems to be a flaky test. If there's not an issue for it already, please file an issue for it

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,137 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ triangle ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor tril_indices(
+    int64_t row, int64_t col, int64_t offset, const TensorOptions& options) {
```
These TensorOptions& aren't being used right now (are you planning on using them later?)

zou3519 replied:
Actually, I realized I don't understand how TensorOptions works, let me look into that a little

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,137 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+Tensor tril_indices(
+    int64_t row, int64_t col, int64_t offset, const TensorOptions& options) {
+  AT_CHECK(row >= 0, "row must be non-negative, got", row);
+  AT_CHECK(col >= 0, "col must be non-negative, got", col);
```
Numpy supports 0 col and/or 0 row (try it out in a python interpreter!). Let's do the same

mrshenli replied:
Hi @zou3519, is this error below an expected behavior or a bug?

```python
>>> import torch
>>> torch.ones(0, 3)
tensor([], size=(0, 3))
>>> torch.ones(0, 3).tril(0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: invalid argument 1: expected a matrix at ../aten/src/TH/generic/THTensorMoreMath.cpp:1255
>>> torch.ones(1, 3).tril(0)
tensor([[1., 0., 0.]])
```

zou3519 replied:
Bug, looks like we don't support 0-sized dims in `torch.tril`. Please file a bug for it in our issue tracker.

zou3519 replied:
A good rule of thumb for operators like this are "does numpy support it?" In this case, numpy did support tril on `np.zeros([0, 3])`

mrshenli replied:
I see, thanks! I will file an issue for it.

mrshenli replied:
Issue added #15016

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,137 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+
+  // Different combinations of row, col, and offset can lead to two cases:
+  //
+  // Case 1 - Trapezoid (Triangle as a special case): row + offset <= col
```
Isn't this `row + offset >= col`?

zou3519 commented on test/test_torch.py:
```diff
@@ -3757,6 +3757,261 @@ def test_tril(self):
+        # test 3 X 3 matrix, offset = -1
+        ti = torch.triu_indices(3, 3, -1)
+        self.assertEqual(
+            torch.tensor([
```
Eh I do like how it's easy to read in its current form though, that would go away if we changed the testing to compare against a comparison function

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,137 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+  // Refer to tril_indices for algorithm design details
+
+  // number of elements in the first column of the triu
+  auto n_first_col = offset < 0 ?
```
It would be nice to deduplicate the indices computation code between triu_indices and tril_indices. `torch.triu_indices(row, col, offset)` should have as many indices as `row * col - torch.tril_indices(row, col, offset+1)

zou3519 commented on test/test_torch.py:
```diff
@@ -3757,6 +3757,261 @@ def test_tril(self):
 
+    def test_tril_indices(self):
+        # test 1 X 1 matrix
+        ti = torch.tril_indices(1, 1)
```
Just curious, does `device=cuda` work right now? `torch.tril_indices(1, 1, device='cuda')`. I'm wondering if our code generation automatically generates that code as well

zou3519 reviewed (commented):
Looks really good!

I left a few minor comments. Action items are:
1) Make sure this works for inputs with row and/or col = 0
2) Add documentation. Try something like https://github.com/pytorch/pytorch/blob/9b2bd284b3da4c2e703ac1d2fe8d1c40dfa01e93/torch/_torch_docs.py#L3281-L3286, you can verify the docstring has been attached with `help(torch.tril_indices)` in an interpreter.
3) Deduplicate the tril_indices and triu_indices code

fmassa commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,137 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+    }
+  });
+
+  return result.to(options.device());
```
question: do we want to say that we support `tril_indices` generation on the GPU, while all it does is actually CPU->GPU copy?

mrshenli replied:
Hi @fmassa, thanks for the comment. I tried to directly pass `device=torch.device('cuda:0')` when I use `at::empty` to create the tensor. The tensor creation works OK,  but it runs into _Segment Fault_ when I try to assign values to elements in that tensor. 

The current implementation gets rid of that problem. I got the idea from [here](https://github.com/pytorch/pytorch/blob/75001458ce4d187f91a4ad98bcb20013a49b86da/aten/src/ATen/native/TensorFactories.cpp#L788-L802), but I am not sure if that is an appropriate implementation here. Do you know what is the correct solution to support GPU? Thanks!

fmassa replied:
Hi,

Good question. The link that @mrshenli sent creates a tensor from data that is already on the GPU, so it kind of makes sense to first get a CPU tensor and then a cpu-gpu copy.

The way of having it natively run on a GPU would require writing a CUDA kernel for it, which fornthis particular case shouldn't be too hard. But I'm not sure know much more efficient it would be, probably quite a bit for large tensors I'd say.
But I wouldn't say it's a blocker for this task, but more of a discussion question.

mrshenli committed:
Address comments:

1. deduplicated indices calculation between tril_indices and
triu_indices.
2. refactored tests to a more concise form.
3. added GPU tests

mrshenli committed:
Added docs for tril_indices and triu_indices

mrshenli committed:
Fix indention

fmassa commented:
Also, Can we return instead a tuple of 1d tensors? I think it's a pity that nonzero returns a single 2d tensor

zou3519 commented:
@fmassa do you think it is more useful as a tuple of 1d tensors?

fmassa commented:
I think yes, because it can be straight away applied as an input to an indexing op, which the 2d tensor needs first to be unbinded. nonzero in numpy returns a tuple of 1d tensors, but I'm not sure if there is a numpy-equivalent of triu_indices

fmassa commented:
Also, if we return a tuple of 1d tensors, might be better to have the 1d tensors to be contiguous, so that the shape of the 2d tensor internally should be 2xn, instead of nx2

mrshenli commented:
Hi @fmassa,

Thanks. Regarding returning a tuple of tensors, I agree that would be more convenient. It requires modifying code generation [template](https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_factories.py#L12). Shall I create another issue to change the template or should it be handled implicitly in this PR?

fmassa commented:
Oh, do we need to change codegen? Don't we have torch.chunk implemented in cpp? Or is it reimplemented in Python?

mrshenli commented:
@fmassa I previously tried to change [native_functions.yaml](https://github.com/pytorch/pytorch/blob/c2e6834d85b6cbc6cccc077e1defa7a8acfbb5ff/aten/src/ATen/native/native_functions.yaml#L2670) to return `(Tensor, Tensor)`, but the generated code in `torch/csrc/autograd/generated/variable_factories.h` always returns `Tensor` as follows:

```c++
inline at::Tensor tril_indices(int64_t row, int64_t col, int64_t offset = 0, const at::TensorOptions & options = {}) {
```

I thought it might be due to the above template, but I am not sure (sorry, I am a pytorch noob)  :)

fmassa commented:
Can you try returning a TensorList or a std::vector<Tensor> instead?

zou3519 commented:
@fmassa numpy does have an equivalent for tril_indices that returns a tuple of 1D tensors instead of one 2D tensor.

mrshenli commented:
Hi @fmassa @zou3519 

It turns out that using `TensorList`/`std::vector<Tensor>` hits the same error due to the generated file:

```
In file included from ../torch/csrc/cuda/Module.cpp:22:0:
../torch/csrc/autograd/generated/variable_factories.h: In function ‘at::Tensor torch::tril_indices(int64_t, int64_t, int64_t, const at::TensorOptions&)’:
../torch/csrc/autograd/generated/variable_factories.h:1589:103: error: conversion from ‘std::vector<at::Tensor>’ to non-scalar type ‘at::Tensor’ requested
   at::Tensor tensor = at::tril_indices(row, col, offset, at::TensorOptions(options).is_variable(false));
```

So, how do we want to proceed from here?

* option 1: Create a new issue to change the template to support multiple return types. Fix the new issue first, and then merge this one.
* option 2: Implicitly address the template issue in this PR.
* option 3: Merge this PR first (after addressing all comments except the template one), and then create a new issue for the template change.

I would vote for option 3.

mrshenli committed:
Address Comments:

1. Return a 2-by-n Tensor, instead of n-by-2
2. Revised docs accordingly

mrshenli commented:
@fmassa @fmassa Do you know how can I configure codegen to allow it to return a tuple of tensors? I saw some existing functions (e.g., `eig`) can do that.

mrshenli committed:
Remove CPU to GPU copy, and stop claiming support GPU

fmassa commented:
Let's just return a 2d tensor for now. It's a pity that it won't be usable as-is for indexing, but well, we can live with that for now maybe?
We can change it to return a tuple when (if) we change `nonzero` as well.

SsnL commented:
In codegen there is a computed flag with name is_tensor_factory or something similar. You can look into it (and maybe special case this one). 

I didn’t read all comments, but shouldn’t returning a tuple really easy if you do something similar to the definition of eig? Or is this also related to that it is considered as a factory method?

Also, I wonder if this will still be considered as factory if you only take in a device rather than the full tensor options.

mrshenli commented:
Hi @SsnL 
Thanks a lot for the suggestions. The `eig` function is defined in `aten/src/ATen/native/LegacyDefinitions.cpp`. I previously thought maybe new features should not go into legacy files (not sure)?

Regarding the `is_tensor_factory` flag, I could not find it. Do you happen to remember the exact name? Thanks!

mrshenli commented:
hmm, I think it might work if we treat it with `gen_variable_type` instead of `gen_variable_factories`. Let me try if I can add `tril_indices` into `gen_variable_factories.py`

mrshenli commented:
And regarding the above flag, I think it might be `is_factory_method`.

zou3519 commented:
@mrshenli if you can make it work, then all the better :)

mrshenli commented:
hmm, `is_factory_method` actually cannot dictate whether a method goes into `gen_variable_factories.py`. The `is_factory_method` flag of the following methods are all set to `False` but still been processed by `gen_variable_factories.py`.

```
_cudnn_init_dropout_state
empty
empty_like
full_like
ones_like
rand_like
randint_like
...
```
Let me try something different.

mrshenli commented:
Unfortunately, forcing `tril_indices` to be processed by `gen_variable_type` breaks some other part of the code:

```
../torch/csrc/jit/generated/register_aten_ops_1.cpp: In lambda function:
../torch/csrc/jit/generated/register_aten_ops_1.cpp:8268:26: error: ‘tril_indices’ is not a member of ‘torch’
           auto result_ = torch::tril_indices((std::move(peek(stack, 0, 6))).toInt(),
                          ^
../torch/csrc/jit/generated/register_aten_ops_1.cpp:8268:26: note: suggested alternatives:
In file included from ../aten/src/ATen/ATen.h:11:0,
                 from ../torch/csrc/jit/attributes.h:7,
                 from ../torch/csrc/jit/ir.h:3,
                 from ../torch/csrc/jit/operator.h:7,
                 from ../torch/csrc/jit/generated/register_aten_ops_1.cpp:1:
aten/src/ATen/Functions.h:5414:41: note:   ‘at::tril_indices’
 static inline std::tuple<Tensor,Tensor> tril_indices(int64_t row, int64_t col, int64_t offset, const TensorOptions & o
ptions) {
```

mrshenli commented:
It seems there is no easy way to return a tuple of Tensors for functions in `TensorFactories.cpp`, but I will be glad to work on that (revise codegen) after this PR. 

@zou3519 @fmassa any comments?

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,141 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+    }
+  });
+
+  return result.transpose(0, 1);
```
For most pytorch ops we guarantee that the result is contiguous. Is there a good reason for transposing this tensor? (How much does it matter to performance?)

zou3519 replied:
Here's the other part of this: if the operation is fast (less than ~6us) the tensor allocation overhead of creating a new tensor for the transpose will dominate

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,141 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+
+  auto n_indices = get_tril_size(row, col, offset);
+
+  // create an empty Tensor with correct soize
```
nit: s/soize/size/g

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,141 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+
+  auto n_indices = row * col - get_tril_size(row, col, offset - 1);
+
+  // create an empty Tensor with correct soize
```
nit: s/soize/size/g

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -5044,6 +5089,51 @@ def parse_kwargs(desc):
+    >>> a = torch.triu_indices(3, 3)
+    >>> a
+    tensor([[0., 0., 0., 1., 1., 2.],
+            [0., 1., 2., 1., 2., 2.]])
```
Can we return LongTensor by default?

zou3519 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -514,6 +514,141 @@ Tensor& range_out(Tensor& result, Scalar start, Scalar end, Scalar step) {
+
+
+Tensor triu_indices(
+    int64_t row, int64_t col, int64_t offset, const TensorOptions& options) {
```
Because we accept TensorOptions& (this is a side effect of this function being a tensor creation method), I think the function in python-land will take the following arguments: `device, requires_grad, layout`. Could you check to see if the function does take these in python land and make sure this code throws an error if anything incorrect is passed in, and then write some tests for this?
The function currently supports `device='cpu'`, `layout=torch.strided`, `requires_grad=True`.

mrshenli replied:
Yes, python-land indeed takes `device, requires_grad, layout`.  Tests are added to make sure we only accept `device=cpu`, and `layout=torch.strided`.

zou3519 reviewed (commented):
Looking great! The main algorithm is good, and the code is a lot cleaner now. I have three last action items:
1) the default return value should probably be a LongTensor (is it? According to the docs you wrote it looks like the default is float)
2) We're transposing the tensor before returning it. Most pytorch functions return contiguous tensors; maybe we should re-think this or make a note in the documentation.
3) Since this is a tensor creation function it can take in **a lot** of extra arguments, like 'device' and 'layout'. We should assert that layout is torch.strided and device is 'cpu' for now (and probably add these arguments to the docs)

fmassa commented:
Let's just return a 2d tensor, like nonzero.
Thanks for looking into it!

mrshenli committed:
Address Comments:

1. Set default dtype to torch.long for both tril_ and triu_indices.
2. Added tests for the above change.
3. Modified docs accordingly.

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2667,6 +2667,10 @@
 - func: tril(Tensor self, int64_t diagonal=0) -> Tensor
   variants: method, function
 
+- func: tril_indices(int64_t row, int64_t col, int64_t offset=0, TensorOptions options=at::kLong) -> Tensor
```
I did not know one could do this, interesting

zou3519 commented on test/test_torch.py:
```diff
@@ -3757,6 +3757,91 @@ def test_tril(self):
         torch.tril(x, out=res2)
         self.assertEqual(res1, res2, 0)
 
+    def _compare_trilu_indices(self, row, col, offset=0, dtype=torch.uint8):
```
Shouldn't the default dtype here be `torch.long`?

zou3519 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -228,7 +228,8 @@ def group_declarations_by_name(declarations, should_bind_fn):
 
 def get_type_default(declaration):
-    if declaration['name'].startswith('randperm'):
+    if declaration['name'].startswith('randperm') or \
```
can you change it so that it is more explicit, i.e. `declaration['name'] == triu_indices or declaration['name'] == tril_indices`?

zou3519 commented on test/test_torch.py:
```diff
@@ -3757,6 +3757,91 @@ def test_tril(self):
+                torch.triu_indices(row, col, offset, dtype=dtype))
+
+    def test_tril_and_triu_indices(self):
+
```
There's not actually a test for the default long behavior (calling tril_indices without a dtype argument), please add one or modify _compare_trilu_indices to not send an argument to torch.triu_indices.

mrshenli replied:
Will tests on line 3825-3833 be sufficient to cover that?

zou3519 replied:
Good point, that is sufficient. My bad I was reading that too fast

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -4983,6 +4983,54 @@ def parse_kwargs(desc):
+where :math:`d_{1}, d_{2}` are the dimensions of the matrix.
+
+Args:
+    row (``int...``): number of rows in the 2-D matrix.
```
What is the "..." here?

mrshenli replied:
Sorry, my bad.

zou3519 reviewed (approved):
Nice work!

Minor nits on testing and documentation but everything else looks great

mrshenli committed:
Address Comments
